### PR TITLE
Recompute upgrade

### PR DESCRIPTION
**Pull request by wuhuachaocoding**
### PR types

New features
### PR changes

Others
### Describe
【Recompute upgrade】:
in the past, there is just one PyLayer implementation of recompute, here , this PR adds other hook implementation of recompute.

【test】

image

【features】
A100 40G fleetx
image

【loss align】
image

### CODE REVIEW

File: python/paddle/distributed/fleet/recompute/recompute.py [hunk 1]
```diff
@@ -383,6 +376,120 @@ def backward(ctx, *args):
...
 
 
+def _recompute_without_reentrant(
+    function, preserve_rng_state=True, *args, **kwargs
+):
+    """
+    recompute without reentrant, that means use hook to implement the recompute function rather than re-entrant autograd.
+    """
+
+    if preserve_rng_state:
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/recompute/recompute.py:384]: User can't use CPU or other XXPU?

File: python/paddle/distributed/fleet/recompute/recompute.py [hunk 2]
```diff
@@ -383,6 +376,120 @@ def backward(ctx, *args):
...
+        amp_level = 'O2'
+    elif tracer._amp_level in (core.AmpLevel.O1, core.AmpLevel.O0):
+        amp_level = 'O1'
+    else:
+        raise ValueError("unsupported amp level: {}".format(tracer._amp_level))
+
+    if tracer._amp_dtype == 'float16':
+        amp_dtype = 'float16'
+    elif tracer._amp_dtype in ('bfloat16', 'float32'):
+        amp_dtype = 'bfloat16'
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/recompute/recompute.py:406]: float32->bfloat16?

File: python/paddle/distributed/fleet/recompute/recompute.py [hunk 3]
```diff
@@ -383,6 +376,120 @@ def backward(ctx, *args):
...
+            raise RuntimeError(
+                "Recompute with RNG perserve is not support current device: {}.".format(
+                    cur_device
+                )
+            )
+        fw_cuda_rng_state = paddle.get_cuda_rng_state()
+        fwd_cuda_rng_state_tracker = (
+            get_rng_state_tracker().get_states_tracker()
+        )
+    tracer = framework._dygraph_tracer()
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/recompute/recompute.py:396]: Are these functions all internal functions with _?

File: python/paddle/distributed/fleet/recompute/recompute.py [hunk 4]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 
 import paddle
+import weakref
 from paddle.fluid import core
```
Comment (gongweibao) [inline python/paddle/distributed/fleet/recompute/recompute.py]: Don't use fluid now.

**Review by gongweibao — CHANGES_REQUESTED** [review]
Modify docs and put the link to comments.

File: python/paddle/distributed/fleet/recompute/recompute.py (hunk 1 above)
Comment (wuhuachaocoding) [reply python/paddle/distributed/fleet/recompute/recompute.py:384]: only support GPU to preserve rng state now.

File: python/paddle/distributed/fleet/recompute/recompute.py (hunk 3 above)
Comment (wuhuachaocoding) [reply python/paddle/distributed/fleet/recompute/recompute.py:396]: yes, and only use this way to get tracer from python, the name of _dygraph_tracer is defined by framewark.

File: python/paddle/distributed/fleet/recompute/recompute.py (hunk 2 above)
Comment (wuhuachaocoding) [reply python/paddle/distributed/fleet/recompute/recompute.py:406]: auto_cast supports two dtypes: float16 and bfloat16.
image

File: python/paddle/distributed/fleet/recompute/recompute.py (hunk 4 above)
Comment (wuhuachaocoding) [reply python/paddle/distributed/fleet/recompute/recompute.py]: DONE

**Review by ForFishes — DISMISSED** [review]
LGTM

File: python/paddle/distributed/fleet/recompute/recompute.py [hunk 5]
```diff
@@ -13,12 +13,16 @@
...
 import paddle
+from paddle import framework
 from paddle.autograd import PyLayer
 from paddle.autograd.py_layer import LegacyPyLayer
-from paddle.fluid import core, framework
-from paddle.fluid.framework import in_dygraph_mode
+from paddle.distributed.fleet.meta_parallel.parallel_layers.random import (
+    get_rng_state_tracker,
+)
+from paddle.framework import core, in_dygraph_mode
```
Comment (sljlp) [inline python/paddle/distributed/fleet/recompute/recompute.py:26]: OK

Comment (wuhuachaocoding) [reply python/paddle/distributed/fleet/recompute/recompute.py:26]: ok

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**